sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -82,8 +108,165 @@ case class GpuIf(
+      gpuFalseExpr: GpuExpression,
+      colTypes: Array[DataType]): GpuColumnVector = {
+
+    def batchFromCv(t: GpuColumnVector, dt: DataType): ColumnarBatch = {
+      withResource(new Table(t.getBase)) { tbl =>
+        GpuColumnVector.from(tbl, Array(dt))
+      }
+    }
+
+    withResource(GpuColumnVector.from(batch)) { tbl =>
+      withResource(pred.getBase.unaryOp(UnaryOp.NOT)) { inverted =>
+        // evaluate true expression against true batch
+        val tt = withResource(filterBatch(tbl, pred.getBase, colTypes)) { trueBatch =>
+          gpuTrueExpr.columnarEval(trueBatch)
+        }
+        withResourceIfAllowed(tt) { _ =>
+          // evaluate false expression against false batch
+          val ff = withResource(filterBatch(tbl, inverted, colTypes)) { falseBatch =>
+            gpuFalseExpr.columnarEval(falseBatch)
+          }
+          withResourceIfAllowed(ff) { _ =>
+            val finalRet = (tt, ff) match {
+              case (t: GpuColumnVector, f: GpuColumnVector) =>
+                withResource(batchFromCv(t, trueExpr.dataType)) { trueTable =>
+                  withResource(batchFromCv(f, falseExpr.dataType)) { falseTable =>
+                    withResource(gather(pred.getBase, trueTable)) { trueValues =>
+                      withResource(gather(inverted, falseTable)) { falseValues =>
+                        pred.getBase.ifElse(trueValues.getColumn(0), falseValues.getColumn(0))
+                      }
+                    }
+                  }
+                }
+              case (t: GpuScalar, f: GpuColumnVector) =>
+                withResource(batchFromCv(f, falseExpr.dataType)) { falseTable =>
+                  withResource(gather(inverted, falseTable)) { falseValues =>
+                    pred.getBase.ifElse(t.getBase, falseValues.getColumn(0))
+                  }
+                }
+              case (t: GpuColumnVector, f: GpuScalar) =>
+                withResource(batchFromCv(t, trueExpr.dataType)) { trueTable =>
+                  withResource(gather(pred.getBase, trueTable)) { trueValues =>
+                    pred.getBase.ifElse(trueValues.getColumn(0), f.getBase)
+                  }
+                }
+              case (_: GpuScalar, _: GpuScalar) =>
+                throw new IllegalStateException(
+                  "scalar expressions can never have side effects")
+            }
+            GpuColumnVector.from(finalRet, dataType)
+          }
+        }
+      }
+    }
+  }
+
+  private def filterBatch(
+      tbl: Table,
+      pred: ColumnVector,
+      colTypes: Array[DataType]): ColumnarBatch = {
+    withResource(tbl.filter(pred)) { filteredData =>
+      GpuColumnVector.from(filteredData, colTypes)
+    }
+  }
+
+  private def boolToInt(cv: ColumnVector): ColumnVector = {
+    withResource(GpuScalar.from(1, DataTypes.IntegerType)) { one =>
+      withResource(GpuScalar.from(0, DataTypes.IntegerType)) { zero =>
+        cv.ifElse(one, zero)
+      }
+    }
```

Comment:
Can this just be `cv.castTo(DType.INT32)`?

Reply:
I tried this and it caused a regression. I will need to look into this more to understand why.

Reply:
Ah, I bet it's because `ifElse` will cause nulls to become 0 whereas casting will not.